In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import operator
import math
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor 
%matplotlib inline

In [74]:
#df = pd.read_csv("/home/hp/Desktop/College/Sem5/DA/Project/zomato-bangalore-restaurants/zomato.csv")
mylist = []

for chunk in  pd.read_csv("C:/Users/sparsha/Documents/SEM 5/DA Project/zomato/newzomato.csv", sep=',', chunksize=20000):
    mylist.append(chunk)

big_data = pd.concat(mylist, axis= 0)
df = big_data

del mylist

Decision Tree Regressor to fill the NA values 

In [75]:
transformed_df = df.copy()
transformed_df['approx_cost(for two people)'].fillna((transformed_df['approx_cost(for two people)'].median()), inplace=True)

In [76]:
transformed_df['new_Cuisine1'], _ = pd.factorize(transformed_df['Cuisine1'], sort=True)
transformed_df['new_Cuisine2'], _ = pd.factorize(transformed_df['Cuisine2'], sort=True)
transformed_df['new_rest1'], _ = pd.factorize(transformed_df['rest1'], sort=True)
transformed_df['new_listed_in'], _ = pd.factorize(transformed_df['listed_in(type)'], sort=True)
transformed_df['new_location'], _ = pd.factorize(transformed_df['location'], sort=True)
transformed_df['new_online_order'], _ = pd.factorize(transformed_df['online_order'], sort=True)
transformed_df['new_book_table'], _ = pd.factorize(transformed_df['book_table'], sort=True)

to_drop = ['Cuisine1',
           'Cuisine2',
           'rest1',
           'listed_in(type)',
           'location',
           'online_order',
           'book_table',
           'name']
transformed_df.drop(to_drop, inplace=True, axis=1)
#transformed_df.dropna(how='any', inplace=True)

In [78]:
nans = lambda df: df[df.isnull().any(axis=1)]
test = nans(transformed_df)

In [79]:
train = transformed_df.copy()
train.dropna(how='any', inplace=True)

In [80]:
train_x = train.loc[:, train.columns != 'rate']
train_y = train.loc[:, train.columns == 'rate']
test_x = test.loc[:, test.columns != 'rate']
test_y = test.loc[:, test.columns == 'rate']

In [81]:
regressor = DecisionTreeRegressor(random_state = 0)

In [82]:
regressor.fit(train_x, train_y)

DecisionTreeRegressor(criterion='mse', max_depth=None, max_features=None,
                      max_leaf_nodes=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      presort=False, random_state=0, splitter='best')

In [83]:
y_pred = list(regressor.predict(test_x))

In [84]:
test_x['rate'] = y_pred

In [85]:
test_x

approx_cost(for two people)  new_Cuisine1  new_Cuisine2  new_rest1  \
72                           150.0            29            29         11   
75                           500.0            15            37          6   
84                           100.0            11            35         11   
90                           500.0            13             5          6   
91                           400.0            35            35          6   
...                            ...           ...           ...        ...   
50902                        400.0            29            29          6   
50933                        900.0            16            16          7   
50968                        900.0            16            16          7   
50971                        600.0            16            16          1   
50972                       2000.0            16            16          7   

       new_listed_in  new_location  new_online_order  new_book_table  rate  
72                 2             1                 0               0   3.1  
75                 2             1                 0               0   4.0  
84                 2             1                 0               0   3.8  
90                 2             1                 0               0   4.1  
91                 2             1                 0               0   3.3  
...              ...           ...               ...             ...   ...  
50902              4             9                 0               0   3.2  
50933              5            89                 0               0   3.4  
50968              6            89                 0               0   3.4  
50971              6            89                 0               0   3.9  
50972              6            89                 0               0   3.8  

[9970 rows x 9 columns]

In [87]:
new_df = pd.concat([train,test_x],sort=True)

In [90]:
train, test = train_test_split(new_df, test_size=0.2)

In [91]:
train_x = train.loc[:, train.columns != 'rate']
train_y = train.loc[:, train.columns == 'rate']
test_x = test.loc[:, test.columns != 'rate']
test_y = test.loc[:, test.columns == 'rate']

In [92]:
regressor = DecisionTreeRegressor(random_state = 0)

In [93]:
regressor.fit(train_x, train_y)

DecisionTreeRegressor(criterion='mse', max_depth=None, max_features=None,
                      max_leaf_nodes=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      presort=False, random_state=0, splitter='best')

In [94]:
y_pred = list(regressor.predict(test_x))

In [95]:
y_pred = np.around(y_pred,decimals=1)
test_y = list(test_y['rate'])
y_pred = list(y_pred)
acc = 0
for i in range(len(y_pred)):
    if(round((y_pred[i] - test_y[i]),1)<0.2):
        acc = acc + 1
print(acc)
print(acc/len(y_pred))

9147
0.897204512015694


Linear Interpolation to fill NA values

In [96]:
transformed_df2 = df.copy()
transformed_df2 = df.interpolate()

In [97]:
transformed_df2['new_Cuisine1'], _ = pd.factorize(transformed_df2['Cuisine1'], sort=True)
transformed_df2['new_Cuisine2'], _ = pd.factorize(transformed_df2['Cuisine2'], sort=True)
transformed_df2['new_rest1'], _ = pd.factorize(transformed_df2['rest1'], sort=True)
transformed_df2['new_listed_in'], _ = pd.factorize(transformed_df2['listed_in(type)'], sort=True)
transformed_df2['new_location'], _ = pd.factorize(transformed_df2['location'], sort=True)
transformed_df2['new_online_order'], _ = pd.factorize(transformed_df2['online_order'], sort=True)
transformed_df2['new_book_table'], _ = pd.factorize(transformed_df2['book_table'], sort=True)

to_drop = ['Cuisine1',
           'Cuisine2',
           'rest1',
           'listed_in(type)',
           'location',
           'online_order',
           'book_table',
           'name']
transformed_df2.drop(to_drop, inplace=True, axis=1)
transformed_df2.dropna(how='any', inplace=True)


In [98]:
train, test = train_test_split(transformed_df2, test_size=0.2)

In [99]:
train_x = train.loc[:, train.columns != 'rate']
train_y = train.loc[:, train.columns == 'rate']
test_x = test.loc[:, test.columns != 'rate']
test_y = test.loc[:, test.columns == 'rate']

In [100]:
regressor = DecisionTreeRegressor(random_state = 0)

In [101]:
regressor.fit(train_x, train_y)

DecisionTreeRegressor(criterion='mse', max_depth=None, max_features=None,
                      max_leaf_nodes=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      presort=False, random_state=0, splitter='best')

In [102]:
y_pred = list(regressor.predict(test_x))

In [103]:
y_pred = np.around(y_pred,decimals=1)
test_y = list(test_y['rate'])
y_pred = list(y_pred)
acc = 0
for i in range(len(y_pred)):
    if(round((y_pred[i] - test_y[i]),1)<0.2):
        acc = acc + 1
print(acc)
print(acc/len(y_pred))

8399
0.8238352133398725


Quadratic interpolation to fill NA values

In [106]:
transformed_df3 = df.copy()
transformed_df3 = df.interpolate(method='polynomial', order=2)

In [107]:
transformed_df3['new_Cuisine1'], _ = pd.factorize(transformed_df3['Cuisine1'], sort=True)
transformed_df3['new_Cuisine2'], _ = pd.factorize(transformed_df3['Cuisine2'], sort=True)
transformed_df3['new_rest1'], _ = pd.factorize(transformed_df3['rest1'], sort=True)
transformed_df3['new_listed_in'], _ = pd.factorize(transformed_df3['listed_in(type)'], sort=True)
transformed_df3['new_location'], _ = pd.factorize(transformed_df3['location'], sort=True)
transformed_df3['new_online_order'], _ = pd.factorize(transformed_df3['online_order'], sort=True)
transformed_df3['new_book_table'], _ = pd.factorize(transformed_df3['book_table'], sort=True)

to_drop = ['Cuisine1',
           'Cuisine2',
           'rest1',
           'listed_in(type)',
           'location',
           'online_order',
           'book_table',
           'name']
transformed_df3.drop(to_drop, inplace=True, axis=1)
transformed_df3.dropna(how='any', inplace=True)


In [108]:
train, test = train_test_split(transformed_df3, test_size=0.2)


In [109]:
train_x = train.loc[:, train.columns != 'rate']
train_y = train.loc[:, train.columns == 'rate']
test_x = test.loc[:, test.columns != 'rate']
test_y = test.loc[:, test.columns == 'rate']

In [110]:
regressor = DecisionTreeRegressor(random_state = 0)

In [111]:
regressor.fit(train_x, train_y)

DecisionTreeRegressor(criterion='mse', max_depth=None, max_features=None,
                      max_leaf_nodes=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      presort=False, random_state=0, splitter='best')

In [112]:
y_pred = list(regressor.predict(test_x))

In [113]:
y_pred = np.around(y_pred,decimals=1)
test_y = list(test_y['rate'])
y_pred = list(y_pred)
acc = 0
for i in range(len(y_pred)):
    if(round((y_pred[i] - test_y[i]),1)<0.2):
        acc = acc + 1
print(acc)
print(acc/len(y_pred))

8235
0.8077488965179009


Decision Tree Regressor with NA values filled with median : 87.49%
Decision Tree Regressor with NA values filled with Decision tree predictor: 89.72%
Decision Tree Regressor with NA values filled with linear interpolation : 82.38%
Decision Tree Regressor with NA values filled with quadratic interpolation : 80.77%